orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
   graph.AddInitializedTensor(proto_data);
 }
 
+Status FindAllConnectedNodes(Graph& graph,
+                             const Node* node,
+                             std::vector<const Node*>& connected_nodes,
+                             std::set<const NodeArg*>& connected_inputs,
+                             std::set<const NodeArg*>& connected_outputs
+                             ) {
+  ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+      node->InputDefs(),
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (graph.IsInputsIncludingInitializers(&node_arg)) {
+          connected_inputs.insert(&node_arg);
+        } else {
+          const Node* producer_node = graph.GetProducerNode(node_arg.Name());
+          if (producer_node == nullptr) {
+            // got nullptr as producer node. This could be because the input is a constant op which will be optimized
+            // away. Print out this information and continue.
+            LOGS_DEFAULT(WARNING) << "Cannot find producer node for node_arg: " << node_arg.Name() << ". Skipping this node.";
```

Comment:
Not essential for this PR, but we should re-visit the different cases here and if there are other situations aside from constant ops.

Reply:
I think another situation could be the producer is constant none-overwritten. I'll add a TODO here to future re-visit.

---
In reply to: [452256763](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452256763) [](ancestors = 452256763)